pkg/apis/operations/v1alpha1/type.go

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright 2022 The KubeEdge Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NodeUpgradeJob is used to upgrade edge node from cloud side.
+// +k8s:openapi-gen=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster
+type NodeUpgradeJob struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired behavior of NodeUpgradeJob.
+	// +optional
+	Spec UpgradeSpec `json:"spec,omitempty"`
+	// Most recently observed status of the Upgrade.
+	// +optional
+	Status Status `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NodeUpgradeJobList is a list of NodeUpgradeJob.
+type NodeUpgradeJobList struct {
+	// Standard type metadata.
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard list metadata.
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// List of NodeUpgradeJobs.
+	Items []NodeUpgradeJob `json:"items"`
+}
+
+// UpgradeSpec is a description of Upgrade.
+type UpgradeSpec struct {
+	// +Required: Version is the EdgeCore version to upgrade.
+	Version string `json:"version,omitempty"`
+	// UpgradeTool is a request to decide use which upgrade tool. If it is empty,
+	// the Upgrade simply use default upgrade tool keadm to do upgrade operation
+	// +optional
+	UpgradeTool string `json:"upgradeTool,omitempty"`
+	// Timeout for the upgrade.
+	// This limits the duration of the upgrade. If not specified, we'll use 300s
+	// +optional
+	TimeoutSeconds *uint32 `json:"timeoutSeconds,omitempty"`
+	// NodeNames is a request to select some specific nodes. If it is non-empty,
+	// the Upgrade simply select these edge nodes to do upgrade operation
+	// +optional
+	NodeNames []string `json:"nodeNames,omitempty"`
+	// LabelSelector is a selector which must be true for the Upgrade to fit on a node.
+	// Selector which must match a node's labels for the Selector to be operated on that node.
+	// +optional
+	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
```

Comment:
```suggestion
	// LabelSelector is a filter to select member clusters by labels.
	// It must match a node's labels for the NodeUpgradeJob to be operated on that node.
	// Please note that sets of NodeNames and LableSelector are ORed.
	// If both NodeNames and LableSelector are empty, no nodes will run the upgrade job.
	// +optional
	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
```

Reply:
now we use webhook to restrict users can only use one(NodeNames or Selector), and cannot set them both.

Reply:
and must set one
```
// we must specify NodeNames or LabelSelector, and we can only specify only one
if len(upgrade.Spec.NodeNames) == 0 && upgrade.Spec.LabelSelector == nil {
    return fmt.Errorf("both NodeNames and LabelSelctor are NOT specified")
}
if len(upgrade.Spec.NodeNames) != 0 && upgrade.Spec.LabelSelector != nil {
    return fmt.Errorf("both NodeNames and LabelSelctor are specified")
}
```